src/cubeb_audiounit.cpp

```
@@ -1285,6 +1286,133 @@ audiounit_configure_input(cubeb_stream * stm)
+      } else {
+        LOG("(%p) Event: kAudioDevicePropertyBufferFrameSize: New buffer size = %d for scope %d", stm,
+            new_buffer_size, inScope);
+      }
+      stm->buffer_size_change_state = true;
+      break;
+    }
+  }
+}
+
+static int
+audiounit_set_buffer_frame_size(cubeb_stream * stm)
+{
+  // Use latency to calculate buffer size
+  uint32_t output_buffer_frames = stm->latency_frames;
+  OSStatus r = AudioUnitAddPropertyListener(stm->output_unit,
+                                            kAudioDevicePropertyBufferFrameSize,
+                                            buffer_size_changed_callback,
+                                            stm);
+  if (r != noErr) {
+    PRINT_ERROR_CODE("AudioUnitAddPropertyListener/output/kAudioDevicePropertyBufferFrameSize", r);
+    return CUBEB_ERROR;
+  }
+
+  stm->buffer_size_change_state = false;
+
+  r = AudioUnitSetProperty(stm->output_unit,
+                           kAudioDevicePropertyBufferFrameSize,
+                           kAudioUnitScope_Input,
+                           AU_OUT_BUS,
+                           &output_buffer_frames,
+                           sizeof(output_buffer_frames));
+  if (r != noErr) {
+    PRINT_ERROR_CODE("AudioUnitSetProperty/output/kAudioDevicePropertyBufferFrameSize", r);
+
+    r = AudioUnitRemovePropertyListenerWithUserData(stm->output_unit,
+                                                    kAudioDevicePropertyBufferFrameSize,
+                                                    buffer_size_changed_callback,
+                                                    stm);
+    if (r != noErr) {
+      PRINT_ERROR_CODE("AudioUnitAddPropertyListener/output/kAudioDevicePropertyBufferFrameSize", r);
+    }
+
+    return CUBEB_ERROR;
+  }
+
+  int count = 0;
+  while (!stm->buffer_size_change_state && count++ < 30) {
+    // Wait for the change to take effect without pausing the "main" thread
```

By "main" thread I meant the thread that executes `cubeb_stream_init`. It's not clear I will change it.

The magic number is an arbitrary value big enough to ensure that there is a failure and there no point to wait any more. Better to return a failure in that case rather than waiting for ever.